### PR TITLE
CFN support for MultiAZ

### PR DESCRIPTION
**Pull request by AkarshD**
The changes provide CFN support for MultiAZ redshift cluster. It allows us to create/modify MultiAZ redshift cluster using the cloudformation template. The changes in this request allow us to use create-cluster, modify-cluster, restore-from-cluster-snapshot and failover-primary-compute APIs for MultiAZ redshift cluster.

We are adding a new boolean parameter "multiAZ" to differentiate between a Single-AZ and MultiAZ cluster and the parameter is used while setting up the CreateClusterRequest, ModifyClusterRequest and RestoreFromClusterSnapshotRequest.

FailoverPrimaryCompute is a new API which will be supported for MultiAZ redshift cluster and can be used with the help of resource-action attribute which also supports pause-cluster and resume-cluster. The cluster has to be a MultiAZ redshift cluster to trigger FailoverPrimaryCompute API. If we try to trigger failover-primary-compute on a single-az cluster, redshift returns the call with an InvalidClusterState exception and a prompt message.

Co-authored-by: Akarsh Dang [dangakar@amazon.com](mailto:dangakar@amazon.com)

### CODE REVIEW

**Review by mashuaihh — APPROVED** [review]

File: aws-redshift-cluster/src/test/java/software/amazon/redshift/cluster/CreateHandlerTest.java [hunk 1]
```diff
@@ -287,4 +289,92 @@ public void testCreateClusterAndEnableLogging() {
...
+                .clusterStatus("available")
+                .clusterAvailabilityStatus("Available")
+                .allowVersionUpgrade(true)
+                .automatedSnapshotRetentionPeriod(0)
+                .encrypted(true)
+                .multiAZ(MULTIAZ_ENABLED)
+                .enhancedVpcRouting(false)
+                .manualSnapshotRetentionPeriod(1)
+                .publiclyAccessible(false)
+                .clusterSecurityGroups(new LinkedList<ClusterSecurityGroupMembership>())
```
Comment (shwetayakkali) [inline aws-redshift-cluster/src/test/java/software/amazon/redshift/cluster/CreateHandlerTest.java]: can we make these as `Collections.emptyList()`

File: aws-redshift-cluster/src/test/java/software/amazon/redshift/cluster/TestUtils.java [hunk 2]
```diff
@@ -42,6 +44,25 @@ public class TestUtils {
             .vpcSecurityGroups(Collections.emptyList())
             .build();
 
+    final static Cluster MULTIAZ_CLUSTER = Cluster.builder()
```
Comment (shwetayakkali) [inline aws-redshift-cluster/src/test/java/software/amazon/redshift/cluster/TestUtils.java:47]: can we create a common helper method in TestUtils to return a multiAZ cluster handy for unit tests ?

**Review by shwetayakkali — CHANGES_REQUESTED** [review]

File: aws-redshift-cluster/src/test/java/software/amazon/redshift/cluster/CreateHandlerTest.java (hunk 1 above)
Comment (AkarshD) [reply aws-redshift-cluster/src/test/java/software/amazon/redshift/cluster/CreateHandlerTest.java]: Sure. I will send an updated PR using collections.emptyList().

File: aws-redshift-cluster/src/test/java/software/amazon/redshift/cluster/TestUtils.java (hunk 2 above)
Comment (AkarshD) [reply aws-redshift-cluster/src/test/java/software/amazon/redshift/cluster/TestUtils.java:47]: I don't think we need a helper method for that. A static variable in utils class also has the same behavior.

Comment (shwetayakkali) [reply aws-redshift-cluster/src/test/java/software/amazon/redshift/cluster/TestUtils.java:47]: I see it's only for a single unit test, should be good

**Review by shwetayakkali — APPROVED** [review]

**Pull request merged**